crates/matrix-sdk-ui/src/timeline/read_receipts.rs

```diff
@@ -513,14 +500,85 @@ impl TimelineInnerMetadata {
             .and_then(|user_map| user_map.get(&receipt_type))
             .cloned()
         {
+            // Since it is in the timeline, it should be the most recent.
             return Some(receipt);
         }
 
-        room.user_receipt(receipt_type.clone(), ReceiptThread::Unthreaded, user_id)
+        let unthreaded_read_receipt = room
+            .user_receipt(receipt_type.clone(), ReceiptThread::Unthreaded, user_id)
+            .await
+            .unwrap_or_else(|e| {
+                error!("Could not get unthreaded user read receipt of type {receipt_type:?}: {e}");
+                None
+            });
+
+        let main_thread_read_receipt = room
+            .user_receipt(receipt_type.clone(), ReceiptThread::Main, user_id)
             .await
             .unwrap_or_else(|e| {
-                error!("Could not get user read receipt of type {receipt_type:?}: {e}");
+                error!("Could not get main thread user read receipt of type {receipt_type:?}: {e}");
                 None
-            })
+            });
+
+        // If we only have one, return it.
+        let Some(unthreaded_receipt) = &unthreaded_read_receipt else {
+            return main_thread_read_receipt;
+        };
+        let Some(main_thread_receipt) = &main_thread_read_receipt else {
+            return unthreaded_read_receipt;
+        };
+
+        match self.compare_receipts(unthreaded_receipt, main_thread_receipt) {
+            ReceiptCmp::First => unthreaded_read_receipt,
+            ReceiptCmp::Second => main_thread_read_receipt,
+            ReceiptCmp::Unknown => {
+                // As a fallback, let's use the unthreaded read receipt, since it's the one
+                // we should be using.
+                unthreaded_read_receipt
+            }
+        }
+    }
+
+    /// Compares two receipts to know which one is more recent.
+    ///
+    /// Returns `None` if it's not possible to know which one is the more
+    /// recent.
+    fn compare_receipts(
+        &self,
+        first: &(OwnedEventId, Receipt),
+        second: &(OwnedEventId, Receipt),
+    ) -> ReceiptCmp {
```

Comment:
For what it's worth, `PartialOrd` would be a great impl for `&(OwnedEventId, Receipt)`, as it's got the right return type `Option<Ordering>` (`None` == `ReceiptCmp::Unknown`). Then we can even get rid of `ReceiptCmp`, which is nice (less cognitive overhead).

If it's not practical to implement `PartialOrd`, could we return `Option<Ordering>` here at least?

Reply:
Also: this function looks like a great candidate for unit testing; would it make sense to add some tests covering all the branches?

Reply:
It is not possible to implement `PartialOrd`, since we need `self` to find out the order. I didn't use `Ordering` because I find that `Less` or `Greater` has no clear meaning when talking about a read receipt. But if you prefer I'll change it.

Regarding the unit tests, sure I can add some but I already added an integration test that covers all branches in the last commit (see codecov).

Reply:
Done